### PR TITLE
[BEAM-11075] Pardo tests for Go SDK

### PR DESCRIPTION
**Pull request by kamilwu**
This PR adds ParDo load tests for Go SDK. The PR also contains some helper functions for measuring and sending test run times to InfluxDB. The results will become visible on Grafana dashboard as soon as the related Jenkins job starts running periodically.

Lastly, the PR introduces a new option: `parallelism`, which can be used to set the degree of Flink's parallelism.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run Load Tests Go ParDo Flink Batch

**Comment by kamilwu** [comment]
Publish Docker Snapshots

**Comment by kamilwu** [comment]
Run Load Tests Go ParDo Flink Batch

**Comment by kamilwu** [comment]
Run Load Tests Go ParDo Flink Batch

**Comment by kamilwu** [comment]
Run Java PreCommit

**Comment by kamilwu** [comment]
R: @lostluck Could you take a look?
cc: @tysonjh

**Comment by lostluck** [comment]
Run Go PostCommit

File: sdks/go/test/load/util.go [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+//
+// It uses a distribution metric which is updated every time new bundle
+// starts or finishes. The processing time can be extracted by calculating
+// the difference of the maximum and the minimum value of the distribution
+// metric.
+type RuntimeMonitor struct{}
+
+// StartBundle updates a distribution metric.
+func (fn *RuntimeMonitor) StartBundle(ctx context.Context, emit func([]byte, []byte)) {
+	runtime.Update(ctx, time.Now().Unix())
```
Comment (lostluck) [inline sdks/go/test/load/util.go:76]: Note, Unix only provides second granularity. If seconds is too coarse (or doesn't match Java and Python), consider using UnixNano() and converting to the correct granularity (millis or micros).

File: sdks/go/test/load/util.go [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+	runtime = beam.NewDistribution(runtimeMetricNamespace, runtimeMetricName)
+)
+
+func init() {
+	beam.RegisterType(reflect.TypeOf((*RuntimeMonitor)(nil)).Elem())
+}
+
+// RuntimeMonitor is a DoFn to record processing time in the pipeline.
+//
+// It uses a distribution metric which is updated every time new bundle
```
Comment (lostluck) [inline sdks/go/test/load/util.go:68]:
```suggestion
// It uses a distribution metric which is updated every time a new bundle
```

**Review by lostluck — APPROVED** [review]
LGTM!
I can only really speak for the the go code, but the groovy and gradle files seem about right too.

File: sdks/go/test/load/util.go (hunk 1 above)
Comment (lostluck) [reply sdks/go/test/load/util.go:76]: (Notionally I think this is the only thing I need confirmation on before merging this)

File: .test-infra/jenkins/CommonTestProperties.groovy [hunk 3]
```diff
@@ -21,17 +21,36 @@
...
+      DIRECT: "DirectRunner",
+      PORTABLE: "PortableRunner",
+    ]
+
+    def goRunnerNames = [
+      DATAFLOW: "dataflow",
+      SPARK: "spark",
+      FLINK: "flink",
+      DIRECT: "direct",
+      PORTABLE: "universal",
```
[the remark this answers is not in the document]
Comment (kamilwu) [reply .test-infra/jenkins/CommonTestProperties.groovy:52]: I think it's worth doing it now. That will make a big part of infra changes unnecessary. I personally prefer the second option, which registers additional alias in the same package.

File: sdks/go/test/load/util.go (hunk 1 above)
Comment (kamilwu) [reply sdks/go/test/load/util.go:76]: I think there's no reason we can't use `UnixNano()`. The database expects the number of seconds, but the fractional part can provide valuable information, especially for small tests.

**Comment by kamilwu** [comment]
@lostluck I accidentally overwrote your commit by a force push (I could have created a "fix" commit...) If you remember what is was about, please add it once again.

**Comment by kamilwu** [comment]
> @lostluck I accidentally overwrote your commit by a force push (I could have created a "fix" commit...) If you remember what is was about, please add it once again.

Nevermind, I've just found it somehow!

**Comment by kamilwu** [comment]
Run RAT PreCommit

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run Load Tests Go ParDo Flink Batch

File: sdks/go/test/load/util.go [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+}
+
+func newLoadTestResult(value float64) loadTestResult {
+	metric := ""
+	if *influxNamespace == "" {
+		metric = runtimeMetricName
+	} else {
+		metric = fmt.Sprintf("%v_%v", *influxNamespace, runtimeMetricName)
+	}
+	return loadTestResult{timestamp: time.Now().Unix(), metric: metric, value: value}
```
Comment (lostluck) [inline sdks/go/test/load/util.go:129]: Should this be UnixNano as well? Or do the published loadTestResults require second granularity?

**Review by lostluck — APPROVED** [review]
One last spot with Unix() and it's not obvious seconds is right to me. Confirm that and we'll merge it in.

**Comment by lostluck** [comment]
Weird note: it looks like "worker" gets swapped with "****" from any of the Go origin printouts in the console logs.

I don't think it actually has anything to do with this PR, as it would break if they were actually replaced anywhere. Just odd.

File: sdks/go/test/load/util.go (hunk 1 above)
Comment (lostluck) [reply sdks/go/test/load/util.go]: SGTM As long as it ends up with the right units for the DB vs Java and Python, it's fine. I'm less familiar so I must ask :).

File: sdks/go/test/load/util.go (hunk 4 above)
Comment (lostluck) [reply sdks/go/test/load/util.go:129]: I'm pretty sure this is correct, re-reading it all.

Comment (kamilwu) [reply sdks/go/test/load/util.go:129]: Yes, this is correct. The test is going to run every 24 hours, so second precision is totally fine. It works the same in Python and Java tests.

**Comment by kamilwu** [comment]
> I don't think it actually has anything to do with this PR, as it would break if they were actually replaced anywhere. Just odd.

I know this problem. Every "worker" gets swapped, because one of defined Jenkins credentials has username "worker". So Jenkins thinks there's a credential leak and tries to de-identify console logs.

**Pull request merged**